### PR TITLE
Support using flutter with specific version

### PR DESCRIPTION
**Pull request by lext-7**
Add a `version` command in flutter_tools to support using flutter with specific version.  

`flutter version`: list a flutter versions.
`flutter version 1.1.5`: use flutter version 1.1.5.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import '../globals.dart';
+import '../runner/flutter_command.dart';
+import '../version.dart';
+
+class VersionCommand extends FlutterCommand {
+  @override
+  final String name = 'version';
+
+  @override
+  final String description = 'Flutter version commands.';
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]: Please describe what the command does.

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        workingDirectory: Cache.flutterRoot);
+    return runResult.toString().split('\n');
+  }
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    final List<String> tags = await getTags();
+    if (argResults.rest.isEmpty) {
+      tags.forEach(print);
+      return null;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]: Return a successful result here

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        <String>['git', 'tag', '-l', 'v*'],
+        workingDirectory: Cache.flutterRoot);
+    return runResult.toString().split('\n');
+  }
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    final List<String> tags = await getTags();
+    if (argResults.rest.isEmpty) {
+      tags.forEach(print);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]: printStatus

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    } catch (e) {
+      throwToolExit(
+          'Unable to checkout version branch for version $version.');
+    }
+
+    final FlutterVersion flutterVersion = FlutterVersion();
+
+    printStatus('Switching Flutter to version ${flutterVersion.frameworkVersion}...');
+
+    int code;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]: Remove: see below

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    // Check for and download any engine and pkg/ updates.
+    // We run the 'flutter' shell script re-entrantly here
+    // so that it will download the updated Dart and so forth
+    // if necessary.
+    printStatus('');
+    printStatus('Downloading engine...');
+    code = await runCommandAndStreamOutput(<String>[
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]:
```suggestion
    int code = await runCommandAndStreamOutput(<String>[
```

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    printStatus('');
+    printStatus(flutterVersion.toString());
+
+    final String projectRoot = findProjectRoot();
+    if (projectRoot != null) {
+      printStatus('');
+      await pubGet(
+          context: PubContext.pubUpgrade,
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]: nit: here and elsewhere, indent by 2 spaces

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      ],
+      workingDirectory: Cache.flutterRoot,
+      allowReentrantFlutter: true,
+    );
+
+    if (code != 0) {
+      throwToolExit(null, exitCode: code);
+    }
+
+    return null;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]: Here and elsewhere for returns
```suggestion
    return return FlutterCommandResult(ExitStatus.success);
```

**Review by jonahwilliams — CHANGES_REQUESTED** [review]
This will need some tests. You can see the examples in the test folder, they will use `testUsingContext` to inject a mock processManager. You can then have this return whatever you want for the various git commands.

This command should also warn the user if they checkout an older version which does not support this command.

**Comment by lext-7** [comment]
@jonahwilliams Tests are added.

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 8]
```diff
@@ -0,0 +1,131 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]:
```suggestion
// Copyright 2019 The Chromium Authors. All rights reserved.
```

File: packages/flutter_tools/lib/src/commands/version.dart [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        workingDirectory: Cache.flutterRoot
+      );
+    } catch (e) {
+      throwToolExit('Unable to checkout version branch for version $version.');
+    }
+
+    final FlutterVersion flutterVersion = FlutterVersion();
+
+    printStatus(
+      'Switching Flutter to version ${flutterVersion.frameworkVersion}${withForce ? ' with force' : ''}...');
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/version.dart]: Move to one line
```suggestion
      printStatus('Switching Flutter to version ${flutterVersion.frameworkVersion}');
```

File: packages/flutter_tools/test/commands/version_test.dart [hunk 10]
```diff
@@ -0,0 +1,121 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands/version_test.dart]:
```suggestion
// Copyright 2019 The Chromium Authors. All rights reserved.
```

**Comment by jonahwilliams** [comment]
Just some minor nits, cc @dnfield for a second pass

**Review by dnfield — CHANGES_REQUESTED** [review]
Sorry for being late to the party on this, but I'm a little skeptical of including this in the Flutter tooling.  I think it's a great idea, and I think it can certainly be useful, but probably as a separate package (perhaps a `pub global` package.

This seems like it would work something like `rvm` or `pyenv` or perhaps even `gradlew` (which we also recently discussed as something that is clearly valuable but doesn't really belong in the tooling itself).  Making it part of the tool runs the risk of having  user downgrade to a version of Flutter that doesn't support this command, and it also runs the risk of some day downgrading to a version that no longer uses the most current method(s) of distributing Flutter.

**Comment by jonahwilliams** [comment]
@dnfield we do already have the flutter channel command, this is fairly similar and I think is okay - provided there are warnings for going back to old versions.

**Comment by dnfield** [comment]
Fair enough.

I'm not sure how we can properly determine the "safe" version for this though.  That won't be determined until after this patch lands, and depending on a variety of factors it's really hard to say what version would be safe to put in there.  Perhaps it could be set to the commit has immediately before this one - although even that is slightly risky, as someone could land something between updating it here and actually merging this.

**Comment by jonahwilliams** [comment]
We could simply push it a bit in the future and err on false-positive

**Comment by lext-7** [comment]
Is there any other problem? Can this pr be merged?

**Review by dnfield — APPROVED** [review]

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
Thanks for this contribution.

In the future please first file a bug to discuss the rationale for a change.
For details see: https://github.com/flutter/flutter/wiki/Tree-hygiene

**Comment by Denesh11** [comment]
can you help me guys..I have a flutter version 1.22.4 ,dart sdk version 2.10.3 but i need dart sdk less than 2.0.0... how to solve this

**Comment by dpedrinha** [comment]
> can you help me guys..I have a flutter version 1.22.4 ,dart sdk version 2.10.3 but i need dart sdk less than 2.0.0... how to solve this

Found your answer @Denesh11?

**Comment by Denesh11** [comment]
> > can you help me guys..I have a flutter version 1.22.4 ,dart sdk version 2.10.3 but i need dart sdk less than 2.0.0... how to solve this
> 
> Found your answer @Denesh11?

yes, I have found a solution.

**Comment by dpedrinha** [comment]
> > > can you help me guys..I have a flutter version 1.22.4 ,dart sdk version 2.10.3 but i need dart sdk less than 2.0.0... how to solve this
> > 
> > 
> > Found your answer @Denesh11?
> 
> yes, I have found a solution.

@Denesh11 do you really think I was just worried about you?? We don't know each other man, I was asking for the solution. But never mind, I found it too.

**Pull request merged**